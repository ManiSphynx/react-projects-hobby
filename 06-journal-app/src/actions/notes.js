import Swal from "sweetalert2";
import {
  collection,
  addDoc,
  doc,
  updateDoc,
  deleteDoc,
} from "firebase/firestore";
import { db } from "../firebase/firebase-config";
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";
import { fileUpload } from "../helpers/fileUpload";

export const startNewNote = () => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;

    const newNote = {
      title: "",
      body: "",
      date: new Date().getTime(),
    };

    try {
      const docRef = await addDoc(
        collection(db, `${uid}`, "journal", "notes"),
        newNote
      );
      dispatch(activeNote(docRef.id, newNote));
      dispatch(addNewNote(docRef.id, newNote));
    } catch (error) {
      console.log(error);
    }
  };
};

export const activeNote = (id, note) => ({
  type: types.notesActive,
  payload: {
    id,
    ...note,
  },
});

export const addNewNote = (id, note) => ({
  type: types.notesAddNew,
  payload: {
    id,
    ...note,
  },
});

export const startLoadingNotes = (uid) => {
  return async (dispatch) => {
    const notes = await loadNotes(uid);
    dispatch(setNotes(notes));
  };
};

export const setNotes = (notes) => ({
  type: types.notesLoad,
  payload: notes,
});

export const startSaveNote = (note) => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;
    if (!note.url) {
      delete note.url;
    }
    const noteToFirestore = { ...note };
    delete noteToFirestore.id;
    const notesRef = doc(db, `${uid}`, "journal", "notes", `${note.id}`);
    await updateDoc(notesRef, noteToFirestore);
    dispatch(refreshNote(note.id, noteToFirestore));
    Swal.fire("Saved", note.title, "success");
  };
};

export const refreshNote = (id, note) => ({
  type: types.notesUpdated,
  payload: {
    id,
    note: {
      id,
      ...note,
    },
  },
});

export const starUploading = (file) => {
  return async (dispatch, getState) => {
    const { active: activeNote } = getState().notes;

    Swal.fire({
      title: "Uploading...",
      text: "Please Wait...",
      allowOutsideClick: false,
      showConfirmButton: false,
      willOpen: () => {
        Swal.showLoading();
      },
    });

    const fileUrl = await fileUpload(file);
    activeNote.url = fileUrl;
    dispatch(startSaveNote(activeNote));

    Swal.close();
  };
};

export const starDeleting = (id) => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;

    const deleteNotes = doc(db, `${uid}`, "journal", "notes", `${id}`);
    await deleteDoc(deleteNotes);

    dispatch(deleteNote(id));
  };
};

export const deleteNote = (id) => ({
  type: types.notesDelete,
  payload: id,
});

export const noteLogout = () => ({
  type: types.notesLogoutCleaning,
});
