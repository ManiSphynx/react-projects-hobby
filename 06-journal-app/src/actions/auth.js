import Swal from "sweetalert2";
import {
  getAuth,
  signInWithPopup,
  createUserWithEmailAndPassword,
  updateProfile,
  signInWithEmailAndPassword,
  signOut,
} from "firebase/auth";
import { googleAuthProvider } from "../firebase/firebase-config";
import { types } from "../types/types";
import { finishLoading, startLoading } from "./ui";
import { noteLogout } from "./notes";

const auth = getAuth();

export const startLoginEmailPassword = (email, password) => {
  return async (dispatch) => {
    dispatch(startLoading());

    return await signInWithEmailAndPassword(auth, email, password)
      .then((userCredential) => {
        const { uid, displayName } = userCredential.user;

        dispatch(login(uid, displayName));
        dispatch(finishLoading());
      })
      .catch((e) => {
        dispatch(finishLoading());
        Swal.fire("Error", e.message, "error");
      });
  };
};

export const startRegisterWithEmailPasswordName = (email, password, name) => {
  return (dispatch) => {
    createUserWithEmailAndPassword(auth, email, password)
      .then(async ({ user }) => {
        await updateProfile(auth.currentUser, {
          displayName: name,
        });
        dispatch(login(user.uid, user.displayName));
      })
      .catch((e) => {
        Swal.fire("Error", e.message, "error");
      });
  };
};

export const startGoogleLogin = () => {
  return (dispatch) => {
    signInWithPopup(auth, googleAuthProvider).then(({ user }) => {
      dispatch(login(user.uid, user.displayName));
    });
  };
};

export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName,
  },
});

export const starLogout = () => {
  return async (dispatch) => {
    await signOut(auth);
    dispatch(logout());
    dispatch(noteLogout());
  };
};

export const logout = () => ({
  type: types.logout,
});
