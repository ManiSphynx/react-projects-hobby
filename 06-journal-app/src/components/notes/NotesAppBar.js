import React, { useRef } from "react";
import { useDispatch, useSelector } from "react-redux";
import { startSaveNote, starUploading } from "../../actions/notes";

export const NotesAppBar = () => {
  const dispatch = useDispatch();
  const { active } = useSelector((state) => state.notes);
  const handleSave = () => {
    dispatch(startSaveNote(active));
  };
  const inpuEl = useRef(null);

  const handlePictureUpload = () => {
    document.querySelector("#fileSelector").click();
  };

  const handleFileChange = (e) => {
    const file = e.target.files[0];
    if (file) {
      dispatch(starUploading(file));
      inpuEl.current.value = "";
    }
  };

  return (
    <div className="notes__appbar">
      <span>28 de agosto 2021</span>

      <input
        id="fileSelector"
        ref={inpuEl}
        type="file"
        name="file"
        style={{ display: "none" }}
        onChange={handleFileChange}
      />

      <div>
        <button className="btn" onClick={handlePictureUpload}>
          Picture
        </button>
        <button className="btn" onClick={handleSave}>
          Save
        </button>
      </div>
    </div>
  );
};
