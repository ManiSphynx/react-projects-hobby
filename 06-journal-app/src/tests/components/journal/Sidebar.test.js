import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { MemoryRouter } from "react-router-dom";
import { Sidebar } from "../../../components/journal/Sidebar";
import { starLogout } from "../../../actions/auth";
import { startNewNote } from "../../../actions/notes";

jest.mock("../../../actions/auth", () => ({
  starLogout: jest.fn(),
}));
jest.mock("../../../actions/notes", () => ({
  startNewNote: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initState = {
  auth: {
    uid: "VICkecLNvBgQBqEkUDvG8WhQQoj2",
    name: "Emanuel",
  },
  notes: {
    notes: [
      {
        id: "1Ejt8OOgzpOemCON2ObG",
        date: 1634317290704,
        title: "Puto mencho",
        url: "https://res.cloudinary.com/djythc5li/image/upload/v1634317308/s4czvcyqzs9xp0ur6elm.jpg",
        body: "chingas a tu madre y me pelas la verga",
      },
    ],
    active: null,
  },
};

let store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={store}>
    <MemoryRouter>
      <Sidebar />
    </MemoryRouter>
  </Provider>
);

beforeEach(() => {
  store = mockStore(initState);
  jest.clearAllMocks();
});

describe("Pruebas en Sidebar", () => {
  test("Sidebar debe de mostrarse correctamente", () => {
    expect(wrapper).toMatchSnapshot();
  });
  test("debe de llamar startLogout", () => {
    wrapper.find(".btn").prop("onClick")();
    expect(starLogout).toHaveBeenCalled();
  });
  test("debe de llamar el startNewNote", () => {
    wrapper.find(".journal__new-entry").prop("onClick")();
    expect(startNewNote).toHaveBeenCalled();
  });
});
