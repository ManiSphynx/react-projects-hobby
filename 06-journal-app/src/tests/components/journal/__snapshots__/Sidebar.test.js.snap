// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en Sidebar Sidebar debe de mostrarse correctamente 1`] = `
<aside
  className="journal__sidebar"
>
  <div
    className="journal__sidebar-navbar"
  >
    <h3
      className="mt-5"
    >
      <i
        className="fa fa-moon"
      />
      <span>
         
        Emanuel
      </span>
    </h3>
    <button
      className="btn"
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
  <div
    className="journal__new-entry"
    onClick={[Function]}
  >
    <i
      className="far fa-calendar-plus fa-5x"
    />
    <p
      className="mt-5"
    >
      New entry
    </p>
  </div>
  <div
    className="journal__entries "
  >
    <div
      className="journal__entry animate__animated animate__fadeIn animate__faster"
      onClick={[Function]}
    >
      <div
        className="journal__entry-picture"
        style={
          Object {
            "backgroundImage": "url(https://res.cloudinary.com/djythc5li/image/upload/v1634317308/s4czvcyqzs9xp0ur6elm.jpg)",
            "backgroundSize": "cover",
          }
        }
      />
      <div
        className="journal__entry-body"
      >
        <p
          className="journal__entry-title"
        >
          Puto mencho
        </p>
        <p
          className="journal__entry-content"
        >
          chingas a tu madre y me pelas la verga
        </p>
      </div>
      <div
        className="journal__entry-date-box"
      >
        <span>
          viernes
        </span>
        <h4>
          15ยบ
        </h4>
      </div>
    </div>
  </div>
</aside>
`;
